akka-stream/src/main/scala/akka/stream/impl/ActorRefBackpressureSinkActor.scala

```diff
@@ -0,0 +1,108 @@
+/**
+ * Copyright (C) 2015 Typesafe Inc. <http://www.typesafe.com>
+ */
+package akka.stream.impl
+
+import java.util
+
+import akka.actor._
+import akka.dispatch.sysmsg.{ DeathWatchNotification, SystemMessage, Watch }
+import akka.stream.ActorMaterializer
+import akka.stream.Attributes.InputBuffer
+import akka.stream.actor.RequestStrategy
+import akka.stream.stage._
+
+import scala.concurrent.Future
+
+/**
+ * INTERNAL API
+ */
+private[akka] class ActorRefBackpressureSinkStage[Out](ref: ActorRef, onInitMessage: Any,
+                                                       ackMessage: Any,
+                                                       onCompleteMessage: Any)
+  extends SinkStage[Out, Unit]("ActorRefBackpressureSink") {
+
+  val maxBuffer = module.attributes.getAttribute(classOf[InputBuffer], InputBuffer(16, 16)).max
+  require(maxBuffer > 0, "Buffer size must be greater than 0")
+
+  override def createLogicAndMaterializedValue: (GraphStageLogic, Unit) = {
+
+    val buffer: util.Deque[Out] = new util.ArrayDeque[Out]()
+    var acknowledgementReceived = false
+    var completeReceived = false
+    var pullRequestIsSent = true
+
+    val logic = new GraphStageLogic(shape) {
+
+      private val callback: AsyncCallbackNoArgs = getAsyncCallback(() ⇒ {
+        if (!buffer.isEmpty) sendData()
+        else acknowledgementReceived = true
+      })
+
+      private val deathWatchCallback: AsyncCallbackNoArgs =
+        getAsyncCallback(() ⇒ completeStage())
+
+      implicit val internalRefCallback = new akka.actor.MinimalActorRef {
+        override def provider: ActorRefProvider =
+          throw new UnsupportedOperationException("ActorRefBackpressureSink MinimalActor does not provide")
+        override def path: ActorPath =
+          ActorMaterializer.downcast(interpreter.materializer).supervisor.path / "ActorRefWithAckMinimalActor"
+        override def !(message: Any)(implicit sender: ActorRef = Actor.noSender): Unit = {
+          message match {
+            case msg if msg == ackMessage ⇒ callback.invoke()
+            case _                        ⇒ //ignore all other messages
+          }
+        }
+        override def sendSystemMessage(message: SystemMessage): Unit = message match {
+          case DeathWatchNotification(_, _, _) ⇒ deathWatchCallback.invoke()
+          case _                               ⇒ //ignore all other messages
+        }
+      }
```

Comment:
Would be great to encapsulate this in a reusable fashion inside `GraphStageLogic`, e.g. exposing it as
~~~scala
val ref = createStageActorRef(msg => /* do stuff with whatever is received, also SysMsg */)
...
ref.watch(otherRef)
~~~